### PR TITLE
Infer config + docs improvements

### PR DESCRIPTION
**Pull request by fmigneault**
@plstcharles 
Presenting to you a `0.5.0-rc` candidate with new `infer` mode.

# summary

uses inference operations original created by `cs-canada` -related branches, but adjusted in order to employ the traditional `thelper` objects

adding more inference mechanisms in the future will become more "plug-and-play"

# details

* add specialized dockerfile with `geo` packages pre-installed (based on top of `base` thelper image)
* many improvements to docker build + changelog (eg: display tag links directly in github! woohoo)
* improve makefile targets (eg: all targets are auto-documented using their comment)
* add generic impl of inference `Tester` session runner (aka `Trainer.eval` but with safeguards)
* add `infer` mode with standard thelper objects for geo SlidingWindow (based of `thelper.data.geo.utils.sliding_window_inference` bare impl)
* example inference config file for the SlidingWindow dataset/raster

# corresponding internal PR

https://www.crim.ca/stash/projects/VISI/repos/thelper/pull-requests/30

### CODE REVIEW

**Comment by plstcharles** [comment]
merci mister, je regarde le tout dès que possible --- probablement en fin de semaine!

File: Makefile [hunk 1]
```diff
@@ -32,41 +39,46 @@ else
...
+# note: use "\#\#" to escape results that would self-match in this target's search definition
+help:	## print this help message (default)
+	@echo "$(_SECTION)=== $(APP_NAME) help ===$(_NORMAL)"
+	@echo "Please use 'make <target>' where <target> is one of:"
+#	@grep -E '^[a-zA-Z_-]+:.*?\#\# .*$$' $(MAKEFILE_LIST) \
+#		| awk 'BEGIN {FS = ":.*?\#\# "}; {printf "    $(_TARGET)%-24s$(_NORMAL) %s\n", $$1, $$2}'
+	@grep -E '\#\#.*$$' "$(APP_ROOT)/$(MAKEFILE_NAME)" \
+		| awk ' BEGIN {FS = "(:|\\-\\-\\-)+.*?\\#\\# "}; \
+			/\--/ {printf "$(_SECTION)%s$(_NORMAL)\n", $$1;} \
+			/:/   {printf "    $(_TARGET)%-24s$(_NORMAL) %s\n", $$1, $$2} \
```
Comment (plstcharles) [inline Makefile:69]: this is pure haxx

File: setup.cfg [hunk 2]
```diff
@@ -1,7 +1,54 @@
 [bumpversion]
-current_version = 0.4.7
+current_version = 0.5.0-rc
 commit = True
 tag = True
+parse = (?P<major>\d+).(?P<minor>\d+).(?P<patch>\d+)(\-(?P<release>.*))?
```
Comment (plstcharles) [inline setup.cfg:5]: niceeee

File: thelper/infer/base.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                raise RuntimeError(f"Invalid call to 'train' using '{tester.__name__}' (Tester)")
+
+            def train_epoch(self, model, epoch, dev, loss, optimizer, loader, metrics, output_path):
+                raise RuntimeError(f"Invalid call to 'train_epoch' using '{tester.__name__}' (Tester)")
+
+        return TesterWrapper
+    return make_tester
+
+
+class Tester(Trainer):
```
Comment (plstcharles) [inline thelper/infer/base.py:44]: me gusta

**Comment by plstcharles** [comment]
Beau travail mec! Tout semble OK de mon côté, j'ai hâte d'avoir un peu plus de vrai temps libre pour me replonger dans tout ça...

File: Makefile (hunk 1 above)
Comment (fmigneault) [reply Makefile:69]: oh yes it is

**Pull request merged**